File: compress-images.mjs
/**
 * SSK Image Compressor
 * Run from the project root: node compress-images.mjs
 *
 * Converts large PNGs/JPEGs to optimised JPEG (quality 85, max 1600px wide)
 * then patches products.ts so image paths point to the new .jpg files.
 */

import sharp from "sharp";
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

const DIRS = [
  "public/images/shop/clothing",
  "public/images/shop/accessories",
  "public/images/branding",
  "public/images/community",
];

const PRODUCTS_TS = "src/lib/products.ts";
const MAX_WIDTH = 1600;
const QUALITY = 85;

let beforeTotal = 0;
let afterTotal = 0;
const renamed = {}; // old web path -> new web path

async function processDir(dir) {
  const full = path.join(__dirname, dir);
  if (!fs.existsSync(full)) {
    console.log(`Skipping ${dir} (not found)`);
    return;
  }

  const files = fs.readdirSync(full);
  for (const fname of files.sort()) {
    const ext = fname.split(".").pop().toLowerCase();
    if (!["png", "jpg", "jpeg"].includes(ext)) continue;

    const src = path.join(full, fname);
    const base = fname.slice(0, fname.lastIndexOf("."));
    const dst = path.join(full, base + ".jpg");

    const beforeSize = fs.statSync(src).size;
    beforeTotal += beforeSize;

    try {
      const meta = await sharp(src).metadata();
      let pipeline = sharp(src);

      if (meta.width > MAX_WIDTH) {
        pipeline = pipeline.resize(MAX_WIDTH, null, { fit: "inside" });
      }

      await pipeline
        .jpeg({ quality: QUALITY, progressive: true, mozjpeg: false })
        .toFile(dst + ".tmp");

      // Atomic replace
      fs.renameSync(dst + ".tmp", dst);

      const afterSize = fs.statSync(dst).size;
      afterTotal += afterSize;
      const savedPct = (((beforeSize - afterSize) / beforeSize) * 100).toFixed(0);
      const status = ext !== "jpg" ? "→ NEW" : "↺ OPT";
      console.log(
        `${status}  ${fname.padEnd(50)}  ${Math.round(beforeSize / 1024).toString().padStart(5)} KB → ${Math.round(afterSize / 1024).toString().padStart(4)} KB  (${savedPct}% saved)`
      );

      if (ext !== "jpg") {
        const relSrc = "/" + dir.replace(/\\/g, "/") + "/" + fname;
        const relDst = "/" + dir.replace(/\\/g, "/") + "/" + base + ".jpg";
        renamed[relSrc] = relDst;
      }
    } catch (err) {
      beforeTotal -= beforeSize; // don't count it in totals
      console.log(`✗ SKIP  ${fname.padEnd(50)}  (${err.message})`);
    }
  }
}

async function main() {
  for (const dir of DIRS) {
    await processDir(dir);
  }

  const mb = (n) => (n / 1024 / 1024).toFixed(1);
  console.log("\n" + "=".repeat(60));
  console.log(
    `TOTAL  ${mb(beforeTotal)} MB  →  ${mb(afterTotal)} MB  (${(((beforeTotal - afterTotal) / beforeTotal) * 100).toFixed(0)}% saved)`
  );
  console.log("=".repeat(60) + "\n");

  // Patch products.ts
  const tsPath = path.join(__dirname, PRODUCTS_TS);
  if (Object.keys(renamed).length > 0 && fs.existsSync(tsPath)) {
    let content = fs.readFileSync(tsPath, "utf-8");
    let patched = content;
    for (const [oldPath, newPath] of Object.entries(renamed)) {
      patched = patched.split(oldPath).join(newPath);
    }
    if (patched !== content) {
      fs.writeFileSync(tsPath, patched, "utf-8");
      console.log(`✓ Patched ${PRODUCTS_TS} — updated ${Object.keys(renamed).length} path(s) to .jpg`);
    } else {
      console.log(`ℹ  ${PRODUCTS_TS} already up to date`);
    }
  }

  console.log("\nDone! Run:  vercel --prod");
  console.log("Once live looks good, delete the old .png files from public/images/");
}

main().catch(console.error);

File: compress-images.py
"""
SSK Image Compressor
Run once from the project root: python compress-images.py

What it does:
  - Converts all product/community PNGs and JPEGs to optimised JPEG (quality 85)
  - Saves them with a .jpg extension alongside the originals
  - Prints a patch for products.ts so you can copy-paste the updated image paths
  - Does NOT delete the originals (safe to re-run)

Requirements: pip install pillow
"""

from PIL import Image
import os, re

DIRS = [
    "public/images/shop/clothing",
    "public/images/shop/accessories",
    "public/images/branding",
    "public/images/community",
]

PRODUCTS_TS = "src/lib/products.ts"

before_total = after_total = 0
renamed: dict[str, str] = {}  # old_path -> new_path (relative to public/)

for d in DIRS:
    if not os.path.isdir(d):
        print(f"Skipping {d} (not found)")
        continue
    for fname in sorted(os.listdir(d)):
        ext = fname.lower().rsplit(".", 1)[-1]
        if ext not in ("png", "jpg", "jpeg"):
            continue

        src = os.path.join(d, fname)
        base = fname.rsplit(".", 1)[0]
        dst = os.path.join(d, base + ".jpg")

        before = os.path.getsize(src)
        before_total += before

        img = Image.open(src).convert("RGB")

        # Cap at 1600px wide — plenty for a product shot displayed at 25vw
        if img.width > 1600:
            ratio = 1600 / img.width
            img = img.resize((1600, int(img.height * ratio)), Image.LANCZOS)

        img.save(dst, "JPEG", quality=85, optimize=True, progressive=True)

        after = os.path.getsize(dst)
        after_total += after

        saved_pct = (before - after) / before * 100
        status = "→ NEW" if not src.endswith(".jpg") or src != dst else "↺ OPT"
        print(f"{status}  {fname:50s}  {before // 1024:5d} KB → {after // 1024:4d} KB  ({saved_pct:.0f}% saved)")

        # Track renames for products.ts patch (only if extension changed)
        if ext != "jpg":
            rel_src = src.replace("public", "").replace("\\", "/")
            rel_dst = dst.replace("public", "").replace("\\", "/")
            renamed[rel_src] = rel_dst

print(f"\n{'='*60}")
print(f"TOTAL  {before_total // 1024 // 1024} MB  →  {after_total // 1024 // 1024} MB  "
      f"({(before_total - after_total) / before_total * 100:.0f}% saved)")
print(f"{'='*60}\n")

# ── Patch products.ts automatically ─────────────────────────────────────────
if renamed and os.path.exists(PRODUCTS_TS):
    with open(PRODUCTS_TS, "r", encoding="utf-8") as f:
        content = f.read()

    patched = content
    for old, new in renamed.items():
        patched = patched.replace(old, new)

    if patched != content:
        with open(PRODUCTS_TS, "w", encoding="utf-8") as f:
            f.write(patched)
        print(f"✓ Patched {PRODUCTS_TS} — updated {len(renamed)} image path(s) to .jpg")
    else:
        print(f"ℹ  No changes needed in {PRODUCTS_TS} (paths already up to date)")
else:
    print("ℹ  No renames or products.ts not found — skipping patch")

print("\nDone! Next steps:")
print("  1. Verify images look correct in the browser (open public/images/shop/.../*.jpg)")
print("  2. Run:  vercel --prod")
print("  3. (Optional) Delete the original .png files once live looks good")
